Make periodic notifications work properly with instant launches

ianmcorvidae opened the pull request:
This is the one place that calls to the `instantlyLaunch` helper function, so it should be the only place that needs to get preferences via bootstrap info to pass to it. This should also mean that anyone who's set up email notifications or not will have them behave correctly with respect to email (in the case where their preference doesn't match the saved launch).

ianmcorvidae commented:
Two things I'm not sure of here:

* should I be doing this within the `useMutation`, or should it go where the mutation is being called from?
* should I encapsulate the "grab the relevant preferences out of preferences" code somewhere in order to reuse it between here and the regular app launch? (if so, where would be good?)

psarando commented on src/serviceFacades/instantlaunches.js:
```diff
@@ -436,6 +441,19 @@ export const instantlyLaunch = ({ instantLaunch, resource, output_dir }) => {
             submission.output_dir = output_dir;
             return submission;
         })
+        .then((submission) => {
+            const notify =
+                preferences?.enableAnalysisEmailNotification || false;
+            submission.notify = notify;
+            const notifyPeriodic =
+                !!preferences?.enablePeriodicEmailNotification;
+            submission.notify_periodic = notifyPeriodic;
+            const periodicPeriod =
+                preferences?.periodicNotificationPeriod || 14400;
+            submission.periodic_period = periodicPeriod;
+
+            return submission;
+        })
```
This works fine as-is, but here's another more compact way to modify this submission JSON:
```suggestion
        .then((submission) => ({
            ...submission,
            notify: !!preferences?.enableAnalysisEmailNotification,
            notify_periodic: !!preferences?.enablePeriodicEmailNotification,
            periodic_period: preferences?.periodicNotificationPeriod || 14400,
        }))
```

ianmcorvidae replied:
This is basically the code I was wondering if I should encapsulate somewhere else (since there is similar `const periodicPeriod = preferences?.periodicNotificationPeriod || 14400;` type code in the app launch code from the last PR) , but condensing it this way is better than doing that, I think.

psarando replied:
I see now. That's a good instinct, but this is probably fine for now. If we need to grab these preferences with these defaults in a 3rd place, then it will definitely be time to refactor 😃

psarando reviewed (changes requested):
> should I encapsulate the "grab the relevant preferences out of preferences" code somewhere in order to reuse it between here and the regular app launch?

I think calling `useBootstrapInfo` from `InstantLaunchButtonWrapper` and passing the preferences into `instantlyLaunch` is fine, unless I'm not understanding the question?

ianmcorvidae commented on src/components/instantlaunches/InstantLaunchButtonWrapper.js:
```diff
@@ -58,46 +60,54 @@ function InstantLaunchButtonWrapper(props) {
-                            analysis.interactive_urls[0]
-                        )}`
-                    );
+    const { mutate: launch } = useMutation(
+        (args) => {
+            args.preferences = bootstrapInfo?.preferences;
+            return instantlyLaunch(args);
+        },
```
Yeah, that's probably better. I had initially not thought of the option of doing it that way for some reason, then realized when I was leaving my comment asking for feedback :sweat_smile:

ianmcorvidae commented:
Thanks for the review, good thoughts. Getting more comfortable with this UI stuff here, slowly, heh. I think I've addressed your comments, @psarando -- did one a little differently (easier to just do `bootstrapInfo.?preferences` right when calling the function, since it's only one place anyway) but I think it should be good.

ianmcorvidae commented on src/components/apps/launch/index.js:
```diff
@@ -80,7 +80,7 @@ const Launch = ({
     );
 
     const preferences = bootstrapInfo?.preferences;
-    const notify = preferences?.enableAnalysisEmailNotification || false;
```
Updated this here to match the suggestion for the other spot that does this. Cleaner, I think.

psarando reviewed (approved):
LGTM 👍 🎉
